File: Java/xunit-test-patterns/src/test/java/_07_picking_test_values/_01_ConstrainedNonDeterminism.java
package _07_picking_test_values;

import autofixture.publicinterface.Any;
import autofixture.publicinterface.InstanceOf;
import org.testng.annotations.Test;

import java.time.LocalDateTime;
import java.util.List;

import static autofixture.publicinterface.InlineGenerators.otherThan;
import static autofixture.publicinterface.InlineGenerators.without;
import static org.assertj.core.api.Assertions.assertThat;

public class _01_ConstrainedNonDeterminism {
    @Test
    public void shouldReturnItsInputWhenItIsNotNull() {
        //GIVEN
        String nonNullValue = Any.string();

        //WHEN
        String result = transform(nonNullValue);

        // THEN
        assertThat(result).isEqualTo(nonNullValue);
    }

    private String transform(String value) {
        if(value == null) {
            return "FAIL";
        } else {
            return value; //TODO make it fail
        }
    }

    @Test
    public void plainGenerators() {
        Any.intValue();
        Any.anonymous(Integer.class);

        Any.localDateTime();
        Any.anonymous(LocalDateTime.class);

        Any.booleanValue();
        Any.anonymous(Boolean.class);

        Any.charValue();
        Any.anonymous(Character.class);

        //...etc

        Any.uri();
        Any.url();
        Any.zonedDateTime();
        //...
    }

    @Test
    public void collections() {
        Any.arrayOf(Integer.class);
        Any.collectionOf(Integer.class);
        Any.sortedMapBetween(String.class, Integer.class);
        //...
    }

    @Test
    public void plainConstrainedGenerators() {
        Any.otherThan(1,2,4,5);
        Any.stringContaining("abc");
        Any.stringOfLength(123);
        //...
    }

    @Test
    public void constrainedGenerators() {
        Integer anInt = Any.anonymous(Integer.class, otherThan(123));
        List<String> stringsWithoutAbc = Any.listOf(String.class, without("abc"));
        Integer[] integers2 = Any.arrayOf(Integer.class, without(123));
    }

    // GENERIC
    @Test
    public void shouldCreateGenericInstances() {
        //WHEN
        MyClass<Integer> anonymous = Any.anonymous(new InstanceOf<MyClass<Integer>>() {});

        //THEN
        assertThat(anonymous.instance).isNotEqualTo(2); //TODO change to isEqualTo()
    }

    class MyClass<T> {
        public T instance;

        public MyClass(T instance) {
            this.instance = instance;
        }
    }

}

File: CSharp/XUnitTestPatterns/XUnitTestPatterns/_06_PickingTestValues/_01_ConstrainedNonDeterminism.cs
using System;
using AutoFixture;
using NUnit.Framework;
using TddXt.AnyRoot;
using TddXt.AnyRoot.Collections;
using TddXt.AnyRoot.Network;
using TddXt.AnyRoot.Numbers;
using TddXt.AnyRoot.Strings;
using TddXt.AnyRoot.Time;
using static TddXt.AnyRoot.Root;

namespace XUnitTestPatterns._06_PickingTestValues
{
  public class _01_ConstrainedNonDeterminism
  {
    [Test]
    public void ShouldReturnItsInputWhenItIsNotNull()
    {
      //GIVEN
      var nonNullValue = Any.String();

      //WHEN
      var result = Transform(nonNullValue);

      //THEN
      Assert.AreEqual(nonNullValue + "a", result);
    }

    private string Transform(string value)
    {
      if (value == null)
      {
        return "FAIL";
      }
      else
      {
        return value + "a"; //TODO make it fail
      }
    }

    [Test]
    public void ShouldReturnItsInputWhenItIsNotNull2()
    {
      //GIVEN
      var fixture = new Fixture();
      var nonNullValue = fixture.Create<string>();

      //WHEN
      var result = Transform(nonNullValue);

      //THEN
      Assert.AreEqual(nonNullValue + "a", result);
    }

    [Test]
    public void PlainGenerators()
    {
      Any.Integer();
      Any.Instance<int>();

      Any.Instance<DateTime>();
      Any.DateTime();

      Any.Boolean();
      Any.Instance<bool>();

      Any.Char();
      Any.Instance<char>();

      //...etc

      Any.Uri();
      Any.UrlString();
      Any.LowerCaseAlphaChar();
      //...
    }

    [Test]
    public void Collections()
    {
      Any.Array<int>();
      Any.List<int>();
      Any.Dictionary<string, int>();
      Any.Dictionary<string, int>();
      //...
    }

    [Test]
    public void PlainConstrainedGenerators()
    {
      Any.OtherThan(1, 2, 4, 5);
      Any.StringContaining("abc");
      Any.String(123);
      //...
    }

    // GENERIC
    [Test]
    public void ShouldCreateGenericInstances()
    {
      //WHEN
      var anonymous = Any.Instance<MyClass<int>>();

      //THEN
      Assert.AreNotEqual(2, anonymous.Instance);
    }


    class MyClass<T>
    {
      public T Instance;

      public MyClass(T instance)
      {
        Instance = instance;
      }
    }

  }
}